File: 20.valid-parentheses.go
package main

func isValid(str string) bool {
	parens := map[rune]rune{')': '(', ']': '[', '}': '{'}
	var stack []rune

	for _, char := range str {
		if char == '(' || char == '[' || char == '{' {
			stack = append(stack, char)
			continue
		}

		if len(stack) > 0 && parens[char] == stack[len(stack)-1] {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}

	return len(stack) == 0
}

File: 20.valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(str)
  stack = []
  brackets = [
    { open: "(", close: ")" },
    { open: "[", close: "]" },
    { open: "{", close: "}" },
  ]

  str.each_char do |chr|
    brackets.each do |bracket|
      stack.push(bracket[:open]) if chr == bracket[:open]

      if chr == bracket[:close]
        return false unless stack.pop == bracket[:open]
      end
    end
  end

  return false unless stack.empty?

  true
end
